### PR TITLE
htlcswitch: relax failure message length check

### PR DESCRIPTION
**Pull request by joostjager**
The spec does not dictate but only recommends a length of 256 bytes. Future tlv extensions may push the failure message length over this limit. 

With this change, receivers can ignore the lengthier extensions without handling it as an unreadable failure. Before long tlv extensions can be used, a sufficiently large number of senders needs to be upgraded. Therefore the sooner this PR is released, the better.

Depends https://github.com/lightningnetwork/lightning-onion/pull/59

### CODE REVIEW

**Comment by joostjager** [comment]
Is there any chance that this can be part of v0.15.2?

**Comment by joostjager** [comment]
Need to decide on how to signal sender capabilities as per https://github.com/lightning/bolts/pull/1021#discussion_r980448134

**Comment by joostjager** [comment]
This PR could be merged regardless of the signaling mechanism though. On its own, I don't think it does any harm and just makes lnd spec compliant.

The same fix in Eclair: https://github.com/ACINQ/eclair/pull/2435

File: lnwire/onion_error.go
Comment (halseth) [inline lnwire/onion_error.go:1227]: is this constant in use anywhere else? If so, we should check whether their use must be changed. If not, maybe it can be deleted.

Comment (joostjager) [reply lnwire/onion_error.go:1227]: Oof, there is a lot. Need to take a look at that.

**Comment by joostjager** [comment]
Short cut coming back at us: https://github.com/lightningnetwork/lnd/pull/3027#discussion_r280013085

**Comment by joostjager** [comment]
I've created two draft PRs that show different ways of getting variable length failure messages though the channel state machine:
* #7030: Implement receiving of a malformed failure in the channel state machine
* #7031: Encode variable length failures so that they can be distinguished from fixed length and malformed failures.

File: lnwire/onion_error.go
Comment (joostjager) [reply lnwire/onion_error.go:1227]: All fixed except for the length check [here](https://github.com/lightningnetwork/lnd/blob/4b827179bbc8742be6726365f669e7db5255c50b/htlcswitch/link.go#L2839). See https://github.com/lightningnetwork/lnd/pull/6913#issuecomment-1276160928

Comment (halseth) [reply lnwire/onion_error.go:1227]: I think perhaps a less disruptive way to become closer to spec compliant is instead of removing this check just require the length to be _at least_ `FailureMessageLength`. 

We are already enforcing _exactly_ `FailureMessageLength` in `lightning_onion` and it seems like we cannot easily just remove the check altogether without opening up a can of works wrt those [converted errors](https://github.com/lightningnetwork/lnd/pull/6913#issuecomment-1276160928). 

If we instead start allowing _larger_ errors than before, we can support the larger errors as we want (and that the spec says we should support), and still we can distinguish the malformed errors by their length, as no valid error message could be `FailureMessageLength +4` including the HMAC.

Thoughts?

Comment (joostjager) [reply lnwire/onion_error.go:1227]: Definitely a nice intermediate step. One thing to consider is that if we [increase the recommended failure length to 1024 bytes](https://github.com/lightning/bolts/pull/1021) eventually, we need to do something extra to also extend converted malformed failures to that length.

Perhaps [when we detect a malformed failure](https://github.com/lightningnetwork/lnd/blob/4b827179bbc8742be6726365f669e7db5255c50b/htlcswitch/link.go#L2841), clip off the padding again, and re-add a longer padding block? Or store the unencrypted failures without padding at all, so that they can be padded right before returning them to the switch.

Comment (joostjager) [reply lnwire/onion_error.go:1227]: If we stick to lengths >= 256, I don't think we need to make any further changes to this PR. Every failure reason != 260 is interpreted as a regular failure already and this would also work for longer reason.

File: docs/release-notes/release-notes-0.16.0.md
Comment (halseth) [inline docs/release-notes/release-notes-0.16.0.md:10]: s/more/closer to being

File: go.mod
Comment (halseth) [inline go.mod:170]: reminder to change this before merging

**Review by halseth — APPROVED** [review]
LGTM 👍

File: docs/release-notes/release-notes-0.16.0.md
Comment (joostjager) [reply docs/release-notes/release-notes-0.16.0.md:10]: Good point, spec compliant is a binary thing. Fixed.

**Comment by joostjager** [comment]
@guggero tagged you because it looked like you had touched the code most in the past. Interested to know whether you want external contributors try to come up with reviewer names, or rather follow a different process?

**Comment by guggero** [comment]
> @guggero tagged you because it looked like you had touched the code most in the past. Interested to know whether you want external contributors try to come up with reviewer names, or rather follow a different process?

I did? Must have been a simple refactor. I lack all the (spec) context to give useful feedback, that's why I asked @Crypt-iQ to review.
No, external contributors (normally) can't request review on PRs. I think with the new sprint progress in place, reviewers are assigned once a PR is selected for a sprint (assuming it is tagged for the next release first).

**Comment by joostjager** [comment]
Updated PR for new linter complaints 😄

File: go.mod
Comment (joostjager) [reply go.mod:170]: now enforced by the linter, no reminder needed 😎

**Comment by Crypt-iQ** [comment]
> Short cut coming back at us: [#3027 (comment)](https://github.com/lightningnetwork/lnd/pull/3027#discussion_r280013085)
> 
> Lightning Labs (@Crypt-iQ perhaps?) - how do we get rid of this?

I haven't written a test, but it seems possible that a regular `UpdateFailHTLC` message can hit this if `Reason` isn't properly formed (though this shouldn't matter as the source would fail during decoding). 

Currently the code assumes it receives a failure size of `2+2+256+32 (hmac)` bytes. And as-is, in the malformed case, the link won't add the hmac so it will be `2+2+256` bytes. In this PR, `EncodeFailure` stays the same, so the number of malformed bytes will stay the same. What we wouldn't be able to do is change `convertedErrorSize` and keep the existing `if len(pd.FailReason) == convertedErrorSize` check. This is because we wouldn't be able to determine whether the failure is malformed w/o the hmac _or_ whether it's a regular failure with a larger failure reason.

I think changing the check might be involved and require persisting something at the forwarding package level to denote that a remote `UpdateFailHTLC` was actually a malformed fail. I think keeping the existing logic _and_ starting to encode larger failures may potentially leak that a given failure message is a malformed failure since it is a specific size. This isn't ideal. If the only concern is the potential privacy leak, we can re-pad the failure to the desired length. If the concern is that it's hacky and leads to issues like https://github.com/lightningnetwork/lnd/issues/7037, I agree (especially if it can be hit via normal, messed up `UpdateFailHTLC`) but it may be tricky to not introduce bugs if we modify the structure of the forwarding packages. It seems like the fix for https://github.com/lightningnetwork/lnd/issues/7037 is to add the same check.

File: lnwire/onion_error.go
Comment (halseth) [inline lnwire/onion_error.go:1227]: I know that it is already checked in `lightning-onion`, but should we also error here if `length < FailureMessageLength`. It was agreed on the spec meeting to just enforce at least 256 lengths (since lnd already disallows anything other than 256).

**Comment by joostjager** [comment]
> I haven't written a test, but it seems possible that a regular `UpdateFailHTLC` message can hit this if `Reason` isn't properly formed (though this shouldn't matter as the source would fail during decoding).

If think this is correct, and I ran the test to verify. If a node X returns a valid 260 bytes failure through `UpdateFailHTLC`, the receiving node Y interprets this as a malformed failure in `processRemoteSettleFails` and adds another hmac. The sender interprets this indeed as an unknown failure originating from the intermediate node Y.

What we should probably do is to validate the reason in `UpdateFailHTLC` and fail the channel if it isn't at least 292 bytes?
 
> I think changing the check might be involved and require persisting something at the forwarding package level to denote that a remote `UpdateFailHTLC` was actually a malformed fail. I think keeping the existing logic _and_ starting to encode larger failures may potentially leak that a given failure message is a malformed failure since it is a specific size. This isn't ideal. If the only concern is the potential privacy leak, we can re-pad the failure to the desired length. If the concern is that it's hacky and leads to issues like #7037, I agree (especially if it can be hit via normal, messed up `UpdateFailHTLC`) but it may be tricky to not introduce bugs if we modify the structure of the forwarding packages. It seems like the fix for #7037 is to add the same check.

I'd be in favor of re-padding and adding the same logic to #7037 and minimize any other changes in this sensitive area. The re-padding doesn't need to be implemented in this PR and can be done when we raise the failure message length to 1024 as proposed in https://github.com/lightning/bolts/pull/1021.

File: lnwire/onion_error.go
Comment (joostjager) [reply lnwire/onion_error.go:1227]: That check would also need to include the pad length, because just the failure message itself is allowed to be short. Not sure if we want to decode the pad length and verify the number of remaining bytes just for a sanity check?

Comment (Crypt-iQ) [reply lnwire/onion_error.go:1227]: I think an extra check can't hurt to make sure that the padding is well-formed if the failure length is < 256

**Review by Crypt-iQ — APPROVED** [review]
linter is failing

**Comment by joostjager** [comment]
Linter is only failing on the go.mod replace: https://github.com/lightningnetwork/lightning-onion/pull/59

File: lnwire/onion_error.go
Comment (joostjager) [reply lnwire/onion_error.go:1227]: Added the check, and now it turns out that the padding in the new longer test vector isn't correct. Will fix.

File: lnwire/onion_error.go
Comment (halseth) [inline lnwire/onion_error.go:1261]: alternatively check `EOF`

File: lnwire/onion_error.go
Comment (halseth) [inline lnwire/onion_error.go:1267]: 👍

File: htlcswitch/link.go
Comment (halseth) [inline htlcswitch/link.go:1858]: If we received such a short reason before this change, what would happen?

We must be sure this is not creating any incompatibilities we didn't have before.

**Review by halseth — COMMENTED** [review]
A question about the new length check enforced in the link, otherwise this LGTM.

File: htlcswitch/link.go
Comment (joostjager) [reply htlcswitch/link.go:1858]: We'd just relay it back to the sender, and risk triggering the `len == 260` rule. I think we can make this change safely as the other implementations acknowledged that they only use 256 bytes failure messages?

File: lnwire/onion_error.go
Comment (joostjager) [reply lnwire/onion_error.go:1227]: Updated https://github.com/lightning/bolts/pull/1021 and test vector in this PR.

File: lnwire/onion_error.go
Comment (joostjager) [reply lnwire/onion_error.go]: shorter, better, fixed.

File: htlcswitch/link.go
Comment (halseth) [reply htlcswitch/link.go:1858]: Assume the channels `Alice<->Bob<->Carol`, and that Alice updates to the version of LND that has this added check (Bob has not updated). Could Carol get Alice to close her channel by maliciously sending a short error that Bob would happily relay?

Comment (joostjager) [reply htlcswitch/link.go:1858]: Hmmm, I see. Good point. Maybe we should just pad the failure here, to at least avoid triggering the length-based malformed failure detection?

Comment (joostjager) [reply htlcswitch/link.go:1858]: Yes, the reason for padding is that we'd at least pass back valid length messages. That opens up the possibility to be stricter about this later down the line.

Comment (Crypt-iQ) [reply htlcswitch/link.go:1858]: What does padding mean here? Only padding (and not adding a mac) to trigger this error
- (https://github.com/lightningnetwork/lightning-onion/blob/b62f49fcadfc013d81c639e196a7f09b428a4dc1/crypto.go#L317-L319)

Comment (joostjager) [reply htlcswitch/link.go:1859]: Updated commit to return a compliant-length array of random bytes.

File: lnwire/onion_error.go
Comment (Crypt-iQ) [inline lnwire/onion_error.go:1266]: `padLength` and `failureLength` are uint16 and they can overflow and still make `totalLength != FailureMessageLength` return false

File: htlcswitch/link.go
Comment (Crypt-iQ) [reply htlcswitch/link.go:1859]: could add a test for this to ensure that if we a node receives < 292, they send backwards a fail message with 292 bytes?

File: lnwire/onion_error.go
Comment (Crypt-iQ) [reply lnwire/onion_error.go:1266]: with the check `totalLength < FailureMessageLength`, the overflow can make a now-valid failure reason fail if the check is true

Comment (Crypt-iQ) [reply lnwire/onion_error.go:1266]: i realize now that it's not possible to overflow given the constraints on message size, still it would be nice to convert to uint32

Comment (joostjager) [reply lnwire/onion_error.go]: Didn't think of this, even if it may not be a problem. Fixed anyway as suggested.

File: htlcswitch/link.go
Comment (joostjager) [reply htlcswitch/link.go:1859]: Added

File: htlcswitch/link_test.go
Comment (Crypt-iQ) [inline htlcswitch/link_test.go:6632]: should pass in `shortReason`

**Review by Crypt-iQ — APPROVED** [review]
lgtm once last test-related comment addressed

File: htlcswitch/link_test.go
Comment (joostjager) [reply htlcswitch/link_test.go:6633]: oops, fixed

File: htlcswitch/link.go
Comment (halseth) [inline htlcswitch/link.go:1873]: For debugging/visibility purposes, could it be an idea to pass back a static "I'm sorry the original error is unreadable"+pad error?

**Review by halseth — CHANGES_REQUESTED** [review]

File: htlcswitch/link.go
Comment (joostjager) [reply htlcswitch/link.go:1873]: Suppose we'd actually handle this outcome differently (so going beyond debugging/visibility), this might motivate an attacker to corrupt failure messages in this exact way.

Also if the origin node receives this message, they won't know which node set it because this is part of the intermediate transform logic. They do know that it must have been an lnd node 😄 

I like the suggestion, but not sure if it is actually good or beneficial. It should never happen with unmodified nodes, because all is set to 256 bytes currently.

Comment (halseth) [reply htlcswitch/link.go:1873]: That's true, not very likely to happen.

**Review by halseth — APPROVED** [review]
LGTM

**Review by carlaKC — COMMENTED** [review]
Just one minor comment then lgtm!

File: lnwire/onion_error.go
Comment (Crypt-iQ) [reply lnwire/onion_error.go:1260]: Don't think so because `nil != io.EOF`

Comment (joostjager) [reply lnwire/onion_error.go:1260]: Indeed, if there was one extra byte, there would be no error.

**Review by carlaKC — APPROVED** [review]

**Comment by Roasbeef** [comment]
@joostjager the dependent PR has been merged, so we can bump the dep here to unblock the CI.

**Comment by joostjager** [comment]
Removed go.mod replace and bumped lightning-onion dep.

File: htlcswitch/failure_test.go
Comment (guggero) [inline htlcswitch/failure_test.go:23]: nit: can use the `//notlint:lll` for the whole test function to avoid needing to append it to each already long line.

File: htlcswitch/failure_test.go
Comment (guggero) [inline htlcswitch/failure_test.go:30]: nit: maybe a file in a `testdata` directory would be better than this ultra long string?

**Review by guggero — APPROVED** [review]

File: htlcswitch/failure_test.go
Comment (joostjager) [reply htlcswitch/failure_test.go:23]: Hmm yes, but then it wouldn't be applied to the code in this function.

File: htlcswitch/failure_test.go
Comment (joostjager) [reply htlcswitch/failure_test.go:30]: I can do that, but the downside is that it isn't as easy to copy/paste/verify with the hex string in the bolt spec. I don't mind the ultra long string, but if you think it's indeed better in a separate binary file, I can do that.

File: htlcswitch/failure_test.go
Comment (guggero) [reply htlcswitch/failure_test.go:23]: I mean like this:
```
// TestLongFailureMessage tests that longer failure messages can be interpreted
// correctly.
//
//nolint:lll
func TestLongFailureMessage(t *testing.T) {
```

Then the comment is valid for the whole block it precedes.

File: htlcswitch/failure_test.go
Comment (guggero) [reply htlcswitch/failure_test.go:30]: The test file can also just be a text file with the hex string.

File: htlcswitch/failure_test.go
Comment (guggero) [inline htlcswitch/failure_test.go:20]: This could at least be wrapped so the long string is on a new line.

File: htlcswitch/failure_test.go
Comment (joostjager) [reply htlcswitch/failure_test.go:23]: Yes, got it. But then the code in that function is also excluded from lll?

File: htlcswitch/failure_test.go
Comment (joostjager) [reply htlcswitch/failure_test.go:20]: Done

File: htlcswitch/failure_test.go
Comment (guggero) [reply htlcswitch/failure_test.go:23]: Ah, I guess so, yes.

Comment (joostjager) [reply htlcswitch/failure_test.go]: Moved all of it into a json file.

File: htlcswitch/failure_test.go
Comment (joostjager) [reply htlcswitch/failure_test.go]: Moved all of it into a json file.

**Pull request merged**
